crates/sui-types/src/effects/effects_v2.rs

```diff
@@ -0,0 +1,288 @@
+                    // create-then-wrapped.
+                }
+                (ObjectIn::NotExist, ObjectOut::ObjectWrite(_, owner), false, false) => {
+                    // unwrapped Move object.
+                    assert!(matches!(
+                        owner,
+                        Owner::ObjectOwner(_) | Owner::AddressOwner(_)
+                    ));
+                }
+                (ObjectIn::NotExist, ObjectOut::ObjectWrite(..), true, false) => {
+                    // created Move object.
+                }
+                (ObjectIn::NotExist, ObjectOut::PackageWrite(_), true, false) => {
+                    // created Move package or user Move package upgrade.
+                }
+                (ObjectIn::Exist(_), ObjectOut::NotExist, false, false) => {
+                    // wrapped.
+                }
+                (ObjectIn::Exist(_), ObjectOut::NotExist, false, true) => {
+                    // deleted.
+                }
+                (
+                    ObjectIn::Exist((old_version, old_digest)),
+                    ObjectOut::ObjectWrite(new_digest, _),
+                    false,
+                    false,
+                ) => {
+                    // mutated.
+                    assert!(old_version.value() < self.lamport_version.value());
+                    assert_ne!(old_digest, new_digest);
+                }
+                (
+                    ObjectIn::Exist((old_version, _)),
+                    ObjectOut::PackageWrite((new_version, _)),
+                    false,
+                    false,
+                ) => {
+                    // system package upgrade.
+                    assert_eq!(old_version.value() + 1, new_version.value());
+                }
+                _ => {
+                    panic!("Impossible object change: {:?}, {:?}", id, change);
+                }
+            }
+        }
+        // Make sure that gas object exists in changed_objects.
+        let (_, owner) = self.gas_object();
+        assert!(matches!(owner, Owner::AddressOwner(_)));
+
+        for (id, _) in &self.unchanged_shared_objects {
+            assert!(unique_ids.insert(*id));
+        }
+    }
+}
+
+#[derive(Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
+struct ObjectChange {
+    // input_state and output_state are the core fields that's required by
+    // the protocol as it tells how an object changes on-chain.
+    /// State of the object in the store prior to this transaction.
+    input_state: ObjectIn,
+    /// State of the object in the store after this transaction.
+    output_state: ObjectOut,
+
+    // id_created and id_deleted are informational fields useful to tell
+    // more comprehensive story of what happened to some objects.
+    /// Whether this object ID is created in this transaction.
+    id_created: bool,
+    /// Whether this object ID is deleted in this transaction.
+    id_deleted: bool,
```

Comment:
If we simply generate a UID in a transaction, will that generate a change like so ?

```
ObjectChange {
    input_state: ObjectIn::NotExist,
    output_state: ObjectOut::NotExist,
    id_created: true,
    id_deleted: false,
}
```

If so, note that currently the limit on the number of new UIDs that can be generated by a single PTB is 2048 -- will it be okay for someone to create transactions that just generates 2048 UIDs, or will that stress the system out? (i.e. should we revisit that limit?)

Also, presumably if a UID is generated and deleted in a single transaction, we don't want it to appear in the effects at all, which implies you can't have both these fields set to true for a given object, so does it make sense to switch this to a tri-state value of some kind? (Created, Deleted, Unchanged)?

Reply:
We won't include it in effects if only an UID is created. Likewise, we won't include it if the UID is created and then deleted.
In the function check_invariant, it lists all allowed combinations of these fields.
I don't have a strong opinion on the tri-state here. Personally I find two bools clearer and easier to understand. If we are concerned about data consistency, we could use a tri-state internally all the way up to effects generation, to minimize the possibility of having the true/true value.

Reply:
> We won't include it in effects if only an UID is created.

`_check_invariants` seems to imply that this is an allowable case (it is the "created and wrapped" case).

No strong opinions on my part regarding tri-state Vs two bools.  I lean towards Tri state to make it clear you can't have both cases simultaneously, but I acknowledge that it's a less standard data type and it will most often be queried as two separate bools).

Reply:
hmm now I am a little confused: if I generate an UID and not delete it in a transaction, is there any other place that I could put it other than the ID of an object? And if so, why do we allow that (i.e. storing UID in a place that's not object ID)?

Reply:
Yes, you're not obliged to use a generated `UID` as the `id` of an object.  You can keep it around to hang dynamic fields off of, that's how `Bag` and `Table` are implemented (they are not objects but they contain `UID`s).

(One thing you cannot do is create a UID in one transaction, stash it for later, and create an object with it -- the ID Leak Verifier requires that only "Fresh" UIDs are used for this, i.e. ones where it can detect that they came from a call to `object::new` and were not previously used as the UID of another object)

Reply:
So it sounds like a UID created but not used will never end up being visible as a key in the live object set? If that's true, then I agree such a UID should be omitted from effects

Reply:
Follow-up question on this one: so it sounds to me like one can create a UID but not use it to create an object, or delete it. Is there a practical way to tell such a case apart from created-then-wrapped?

Reply:
as discussed elsewhere, I agree there's no practical way and even if there was, someone can wrap an object and never unwrap it, so its unclear why we would even need to distinguish between these cases.